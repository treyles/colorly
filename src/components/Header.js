import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { CSSTransitionGroup } from 'react-transition-group';
import { auth } from '../utils/base';
import Icon from '../utils/Icon';
import ClickOutside from '../utils/ClickOutside';
import { closeNewUserDialog } from '../actions';

export class Header extends React.Component {
  static handleSignOutClick() {
    auth.signOut().then(() => {
      localStorage.setItem('authenticated', false);
    });
  }

  constructor(props) {
    super(props);
    this.state = {
      profileOpen: false
    };

    this.handleDialogToggleClick = this.handleDialogToggleClick.bind(this);
    this.handleAddPaletteClick = this.handleAddPaletteClick.bind(this);
    this.closeDialog = this.closeDialog.bind(this);
  }

  handleDialogToggleClick() {
    this.setState({
      profileOpen: !this.state.profileOpen
    });
  }

  handleAddPaletteClick() {
    if (this.props.isNewUser) {
      this.props.closeNewUserDialog();
    }

    this.props.history.push('/palettebuild');
  }

  closeDialog() {
    this.setState({
      profileOpen: false
    });
  }

  render() {
    const { profileOpen } = this.state;
    const { userPhoto, userName, library, animateHeader } = this.props;

    const profileDialog = (
      <div className="profile-dialog">
        <div className="palette-count">
          <span className="count">{library.length}</span>
          {library.length !== 1 ? 'PALETTES' : 'PALETTE'}
        </div>
        <button className="sign-out" onClick={Header.handleSignOutClick}>
          SIGN OUT
        </button>
      </div>
    );

    const headerLogo = (
      <div className="header-logo">
        <h2>Colorly</h2>
        <span>
          <Icon icon="logo" />
        </span>
      </div>
    );

    return (
      <header>
        <div
          className="header-top"
          style={{ backgroundColor: `${animateHeader ? '#fff' : ''}` }}
        >
          <div className="profile">
            <div className="profile-image">
              {userPhoto && <img src={userPhoto} alt="avatar" />}
            </div>
            <button onClick={this.handleDialogToggleClick}>
              {userName}
            </button>
            <button onClick={this.handleDialogToggleClick}>
              <Icon icon="down" />
            </button>
            <ClickOutside
              elementIsOpen={profileOpen}
              onRequestClose={this.closeDialog}
            >
              {profileDialog}
            </ClickOutside>
          </div>
          <button className="add" onClick={this.handleAddPaletteClick}>
            ADD PALETTE
          </button>
        </div>
        <CSSTransitionGroup
          transitionName="header-logo"
          transitionEnterTimeout={300}
          transitionLeaveTimeout={300}
        >
          {!animateHeader && headerLogo}
        </CSSTransitionGroup>
      </header>
    );
  }
}

Header.defaultProps = {
  userName: '',
  userPhoto: null
};

Header.propTypes = {
  history: PropTypes.object.isRequired, // eslint-disable-line
  animateHeader: PropTypes.bool.isRequired,
  closeNewUserDialog: PropTypes.func.isRequired,
  library: PropTypes.arrayOf(PropTypes.object).isRequired,
  userPhoto: PropTypes.string,
  userName: PropTypes.string,
  isNewUser: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  library: state.data.library,
  userPhoto: state.user.currentUser.photoURL,
  userName: state.user.currentUser.displayName,
  isNewUser: state.user.isNewUser
});

export default withRouter(
  connect(mapStateToProps, { closeNewUserDialog })(Header)
);
