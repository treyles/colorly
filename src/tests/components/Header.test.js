import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../components/Header';

describe('Header', () => {
  let component;
  const props = {
    isNewUser: false,
    closeNewUserDialog: jest.fn(),
    currentUser: true,
    library: [],
    animateHeader: false,
    history: { push: jest.fn() }
  };

  beforeEach(() => {
    component = shallow(<Header {...props} />);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('shows number of cards in library', () => {
    component.setProps({
      library: [{}, {}]
    });

    expect(component.find('.count').text()).toEqual('2');
  });

  it('changes profileOpen state to true dialog name button is clicked', () => {
    const profileButton = component.find('button').at(0);

    // default state
    expect(component.state().profileOpen).toEqual(false);
    profileButton.simulate('click');
    expect(component.state().profileOpen).toEqual(true);
  });

  it('calls "handleAddPaletteClick" on ADD button click', () => {
    const spy = jest
      .spyOn(Header.prototype, 'handleAddPaletteClick')
      .mockImplementation(jest.fn());
    const wrapper = shallow(<Header {...props} />);
    wrapper.find('.add').simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('calls "handleSignOutClick" on SIGN OUT button click', () => {
    const spy = jest
      .spyOn(Header.prototype, 'handleSignOutClick')
      .mockImplementation(jest.fn());
    const wrapper = shallow(<Header {...props} />);
    wrapper.find('.sign-out').simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
