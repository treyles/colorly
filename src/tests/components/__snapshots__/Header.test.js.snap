// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    animateHeader={false}
    closeNewUserDialog={[MockFunction]}
    currentUser={true}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    isNewUser={false}
    library={Array []}
    userName=""
    userPhoto={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="header-top"
          style={
            Object {
              "backgroundColor": "",
            }
          }
        >
          <div
            className="profile"
          >
            <div
              className="profile-image"
            />
            <button
              onClick={[Function]}
            >
              
            </button>
            <button
              onClick={[Function]}
            >
              <Icon
                icon="down"
              />
            </button>
            <ClickOutside
              elementIsOpen={false}
              onRequestClose={[Function]}
            >
              <div
                className="profile-dialog"
              >
                <div
                  className="palette-count"
                >
                  <span
                    className="count"
                  >
                    0
                  </span>
                  PALETTES
                </div>
                <button
                  className="sign-out"
                  onClick={[Function]}
                >
                  SIGN OUT
                </button>
              </div>
            </ClickOutside>
          </div>
          <button
            className="add"
            onClick={[Function]}
          >
            ADD PALETTE
          </button>
        </div>,
        <CSSTransitionGroup
          transitionAppear={false}
          transitionEnter={true}
          transitionEnterTimeout={300}
          transitionLeave={true}
          transitionLeaveTimeout={300}
          transitionName="header-logo"
        >
          <div
            className="header-logo"
          >
            <h2>
              Colorly
            </h2>
            <span>
              <Icon
                icon="logo"
              />
            </span>
          </div>
        </CSSTransitionGroup>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="profile"
            >
              <div
                className="profile-image"
              />
              <button
                onClick={[Function]}
              >
                
              </button>
              <button
                onClick={[Function]}
              >
                <Icon
                  icon="down"
                />
              </button>
              <ClickOutside
                elementIsOpen={false}
                onRequestClose={[Function]}
              >
                <div
                  className="profile-dialog"
                >
                  <div
                    className="palette-count"
                  >
                    <span
                      className="count"
                    >
                      0
                    </span>
                    PALETTES
                  </div>
                  <button
                    className="sign-out"
                    onClick={[Function]}
                  >
                    SIGN OUT
                  </button>
                </div>
              </ClickOutside>
            </div>,
            <button
              className="add"
              onClick={[Function]}
            >
              ADD PALETTE
            </button>,
          ],
          "className": "header-top",
          "style": Object {
            "backgroundColor": "",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="profile-image"
                />,
                <button
                  onClick={[Function]}
                >
                  
                </button>,
                <button
                  onClick={[Function]}
                >
                  <Icon
                    icon="down"
                  />
                </button>,
                <ClickOutside
                  elementIsOpen={false}
                  onRequestClose={[Function]}
                >
                  <div
                    className="profile-dialog"
                  >
                    <div
                      className="palette-count"
                    >
                      <span
                        className="count"
                      >
                        0
                      </span>
                      PALETTES
                    </div>
                    <button
                      className="sign-out"
                      onClick={[Function]}
                    >
                      SIGN OUT
                    </button>
                  </div>
                </ClickOutside>,
              ],
              "className": "profile",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": null,
                  "className": "profile-image",
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Icon
                    icon="down"
                  />,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "icon": "down",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <div
                    className="profile-dialog"
                  >
                    <div
                      className="palette-count"
                    >
                      <span
                        className="count"
                      >
                        0
                      </span>
                      PALETTES
                    </div>
                    <button
                      className="sign-out"
                      onClick={[Function]}
                    >
                      SIGN OUT
                    </button>
                  </div>,
                  "elementIsOpen": false,
                  "onRequestClose": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <div
                        className="palette-count"
                      >
                        <span
                          className="count"
                        >
                          0
                        </span>
                        PALETTES
                      </div>,
                      <button
                        className="sign-out"
                        onClick={[Function]}
                      >
                        SIGN OUT
                      </button>,
                    ],
                    "className": "profile-dialog",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": Array [
                          <span
                            className="count"
                          >
                            0
                          </span>,
                          "PALETTES",
                        ],
                        "className": "palette-count",
                      },
                      "ref": null,
                      "rendered": Array [
                        Object {
                          "instance": null,
                          "key": undefined,
                          "nodeType": "host",
                          "props": Object {
                            "children": 0,
                            "className": "count",
                          },
                          "ref": null,
                          "rendered": 0,
                          "type": "span",
                        },
                        "PALETTES",
                      ],
                      "type": "div",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "SIGN OUT",
                        "className": "sign-out",
                        "onClick": [Function],
                      },
                      "ref": null,
                      "rendered": "SIGN OUT",
                      "type": "button",
                    },
                  ],
                  "type": "div",
                },
                "type": [Function],
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "ADD PALETTE",
              "className": "add",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "ADD PALETTE",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            className="header-logo"
          >
            <h2>
              Colorly
            </h2>
            <span>
              <Icon
                icon="logo"
              />
            </span>
          </div>,
          "transitionAppear": false,
          "transitionEnter": true,
          "transitionEnterTimeout": 300,
          "transitionLeave": true,
          "transitionLeaveTimeout": 300,
          "transitionName": "header-logo",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Colorly
              </h2>,
              <span>
                <Icon
                  icon="logo"
                />
              </span>,
            ],
            "className": "header-logo",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Colorly",
              },
              "ref": null,
              "rendered": "Colorly",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Icon
                  icon="logo"
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "icon": "logo",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "span",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": "header",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="header-top"
            style={
              Object {
                "backgroundColor": "",
              }
            }
          >
            <div
              className="profile"
            >
              <div
                className="profile-image"
              />
              <button
                onClick={[Function]}
              >
                
              </button>
              <button
                onClick={[Function]}
              >
                <Icon
                  icon="down"
                />
              </button>
              <ClickOutside
                elementIsOpen={false}
                onRequestClose={[Function]}
              >
                <div
                  className="profile-dialog"
                >
                  <div
                    className="palette-count"
                  >
                    <span
                      className="count"
                    >
                      0
                    </span>
                    PALETTES
                  </div>
                  <button
                    className="sign-out"
                    onClick={[Function]}
                  >
                    SIGN OUT
                  </button>
                </div>
              </ClickOutside>
            </div>
            <button
              className="add"
              onClick={[Function]}
            >
              ADD PALETTE
            </button>
          </div>,
          <CSSTransitionGroup
            transitionAppear={false}
            transitionEnter={true}
            transitionEnterTimeout={300}
            transitionLeave={true}
            transitionLeaveTimeout={300}
            transitionName="header-logo"
          >
            <div
              className="header-logo"
            >
              <h2>
                Colorly
              </h2>
              <span>
                <Icon
                  icon="logo"
                />
              </span>
            </div>
          </CSSTransitionGroup>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="profile"
              >
                <div
                  className="profile-image"
                />
                <button
                  onClick={[Function]}
                >
                  
                </button>
                <button
                  onClick={[Function]}
                >
                  <Icon
                    icon="down"
                  />
                </button>
                <ClickOutside
                  elementIsOpen={false}
                  onRequestClose={[Function]}
                >
                  <div
                    className="profile-dialog"
                  >
                    <div
                      className="palette-count"
                    >
                      <span
                        className="count"
                      >
                        0
                      </span>
                      PALETTES
                    </div>
                    <button
                      className="sign-out"
                      onClick={[Function]}
                    >
                      SIGN OUT
                    </button>
                  </div>
                </ClickOutside>
              </div>,
              <button
                className="add"
                onClick={[Function]}
              >
                ADD PALETTE
              </button>,
            ],
            "className": "header-top",
            "style": Object {
              "backgroundColor": "",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="profile-image"
                  />,
                  <button
                    onClick={[Function]}
                  >
                    
                  </button>,
                  <button
                    onClick={[Function]}
                  >
                    <Icon
                      icon="down"
                    />
                  </button>,
                  <ClickOutside
                    elementIsOpen={false}
                    onRequestClose={[Function]}
                  >
                    <div
                      className="profile-dialog"
                    >
                      <div
                        className="palette-count"
                      >
                        <span
                          className="count"
                        >
                          0
                        </span>
                        PALETTES
                      </div>
                      <button
                        className="sign-out"
                        onClick={[Function]}
                      >
                        SIGN OUT
                      </button>
                    </div>
                  </ClickOutside>,
                ],
                "className": "profile",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": null,
                    "className": "profile-image",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "",
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Icon
                      icon="down"
                    />,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "icon": "down",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": <div
                      className="profile-dialog"
                    >
                      <div
                        className="palette-count"
                      >
                        <span
                          className="count"
                        >
                          0
                        </span>
                        PALETTES
                      </div>
                      <button
                        className="sign-out"
                        onClick={[Function]}
                      >
                        SIGN OUT
                      </button>
                    </div>,
                    "elementIsOpen": false,
                    "onRequestClose": [Function],
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        <div
                          className="palette-count"
                        >
                          <span
                            className="count"
                          >
                            0
                          </span>
                          PALETTES
                        </div>,
                        <button
                          className="sign-out"
                          onClick={[Function]}
                        >
                          SIGN OUT
                        </button>,
                      ],
                      "className": "profile-dialog",
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": Array [
                            <span
                              className="count"
                            >
                              0
                            </span>,
                            "PALETTES",
                          ],
                          "className": "palette-count",
                        },
                        "ref": null,
                        "rendered": Array [
                          Object {
                            "instance": null,
                            "key": undefined,
                            "nodeType": "host",
                            "props": Object {
                              "children": 0,
                              "className": "count",
                            },
                            "ref": null,
                            "rendered": 0,
                            "type": "span",
                          },
                          "PALETTES",
                        ],
                        "type": "div",
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "SIGN OUT",
                          "className": "sign-out",
                          "onClick": [Function],
                        },
                        "ref": null,
                        "rendered": "SIGN OUT",
                        "type": "button",
                      },
                    ],
                    "type": "div",
                  },
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ADD PALETTE",
                "className": "add",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "ADD PALETTE",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <div
              className="header-logo"
            >
              <h2>
                Colorly
              </h2>
              <span>
                <Icon
                  icon="logo"
                />
              </span>
            </div>,
            "transitionAppear": false,
            "transitionEnter": true,
            "transitionEnterTimeout": 300,
            "transitionLeave": true,
            "transitionLeaveTimeout": 300,
            "transitionName": "header-logo",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h2>
                  Colorly
                </h2>,
                <span>
                  <Icon
                    icon="logo"
                  />
                </span>,
              ],
              "className": "header-logo",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Colorly",
                },
                "ref": null,
                "rendered": "Colorly",
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Icon
                    icon="logo"
                  />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "icon": "logo",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "span",
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": "header",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
